service/s3/s3crypto/envelope.go

```diff
@@ -0,0 +1,137 @@
+	} else if value = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key"); value != "" {
+		return getV1Envelope(r)
+	} else {
+		return getFromInstructionFile(input, r)
+	}
+}
+
+// TODO:
+// Need MD5 and SHA256 io.Reader
+// Need to redo all io.Reader interface for encrypt and decrypt
+// Implement KMS and Private Key RSA
+func getV2Envelope(r *request.Request) (*Envelope, error) {
+	env := &Envelope{}
+	env.CipherKey = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key-v2")
+	env.IV = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-iv")
+	env.MatDesc = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-matdesc")
+	env.WrapAlg = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-wrap-alg")
+	env.CEKAlg = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-cek-alg")
+	env.TagLen = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-tag-len")
+	env.UnencryptedMD5 = r.HTTPResponse.Header.Get("x-amz-unencrypted-content-md5")
+	env.UnencryptedContentLen = r.HTTPResponse.Header.Get("x-amz-unencrypted-content-length")
+	env.version = 2
+	return env, nil
+}
+
+func getV1Envelope(r *request.Request) (*Envelope, error) {
+	env := &Envelope{}
+	env.CipherKey = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-key")
+	env.IV = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-iv")
+	env.MatDesc = r.HTTPResponse.Header.Get("x-amz-meta-x-amz-matdesc")
+	env.version = 1
+	return env, nil
+}
+
+func getFromInstructionFile(input *s3.GetObjectInput, r *request.Request) (*Envelope, error) {
+	return &Envelope{}, nil
+}
+
+// SaveStrategy is how the data's metadata wants to be saved
+type SaveStrategy interface {
+	Save(Envelope, *s3.PutObjectInput) error
+}
+
+// default s3 key
+const instructionKey = ".instruction"
+
+type s3SaveStrategy struct {
+	client                *s3.S3
+	InstructionFileSuffix *string
+}
+
+// NewS3SaveStrategy returns a new strategy that will save to S3
+func NewS3SaveStrategy(p client.ConfigProvider, suffix *string) SaveStrategy {
+	return &s3SaveStrategy{s3.New(p), suffix}
+}
+
+// Save will save the envelope contents to s3.
+// TODO: Reimplement
+func (strat *s3SaveStrategy) Save(env Envelope, input *s3.PutObjectInput) error {
+	b, err := json.Marshal(env)
+	if err != nil {
+		return err
+	}
+
+	instInput := s3.PutObjectInput{
+		Bucket: input.Bucket,
+		Body:   bytes.NewReader(b),
+	}
+
+	if strat.InstructionFileSuffix == nil {
+		instInput.Key = aws.String(*input.Key + instructionKey)
+	} else {
+		instInput.Key = aws.String(*input.Key + instructionKey + "-" + *strat.InstructionFileSuffix)
+	}
+
+	_, err = strat.client.PutObject(&instInput)
+	return err
+}
+
+type headerSaveStrategy struct{}
+
+// NewHeaderSaveStrategy returns a new strategy that will save to the metadata
+func NewHeaderSaveStrategy() SaveStrategy {
+	return &headerSaveStrategy{}
+}
+
+// Save will save the envelope to the request's header.
+func (strat *headerSaveStrategy) Save(env Envelope, input *s3.PutObjectInput) error {
+	if input.Metadata == nil {
+		input.Metadata = make(map[string]*string)
```

Comment:
Nit, Both styles are file, but preferable to only use make when needed. `input.Matadata = map[string]*string{}` can be used here.